protocol/rpcconsumer/rpcconsumer_server.go

```diff
@@ -186,83 +188,166 @@ func (rpccs *RPCConsumerServer) sendRelayToProvider(
 	if err != nil {
 		return relayResult, err
 	}
-	relayResult.Request = relayRequest
-	endpointClient := *singleConsumerSession.Endpoint.Client
 
-	if isSubscription {
-		return rpccs.relaySubscriptionInner(ctx, endpointClient, singleConsumerSession, relayResult)
+	type relayResponse struct {
+		relayResult *lavaprotocol.RelayResult
+		err         error
 	}
 
-	// try using cache before sending relay
-	var reply *pairingtypes.RelayReply
+	// Make a channel for all providers to send responses
+	responses := make(chan *relayResponse, len(sessions))
 
-	reply, err = rpccs.cache.GetEntry(ctx, relayRequest, chainMessage.GetInterface().Interface, nil, chainID, false) // caching in the portal doesn't care about hashes, and we don't have data on finalization yet
-	if err == nil && reply != nil {
-		// Info was fetched from cache, so we don't need to change the state
-		// so we can return here, no need to update anything and calculate as this info was fetched from the cache
-		relayResult.Reply = reply
-		lavaprotocol.UpdateRequestedBlock(relayResult.Request.RelayData, reply) // update relay request requestedBlock to the provided one in case it was arbitrary
-		err = rpccs.consumerSessionManager.OnSessionUnUsed(singleConsumerSession)
-		return relayResult, err
-	}
+	// Iterate over the sessions map
+	for providerPublicAddress, sessionInfo := range sessions {
+		// Launch a separate goroutine for each session
+		go func(providerPublicAddress string, sessionInfo *lavasession.SessionInfo) {
+			goroutineCtx, goroutineCtxCancel := context.WithCancel(context.Background())
+			defer goroutineCtxCancel()
 
-	// cache failed, move on to regular relay
-	if performance.NotConnectedError.Is(err) {
-		utils.LavaFormatError("cache not connected", err)
-	}
+			localRelayResult := &lavaprotocol.RelayResult{ProviderAddress: providerPublicAddress, Finalized: false}
+			localRelayRequestData := *relayRequestData
 
-	extraRelayTimeout := time.Duration(0)
-	if chainMessage.GetInterface().Category.HangingApi {
-		_, extraRelayTimeout, _, _ = rpccs.chainParser.ChainBlockStats()
-	}
-	relayTimeout := extraRelayTimeout + common.GetTimePerCu(singleConsumerSession.LatestRelayCu) + common.AverageWorldLatency
-	relayResult, relayLatency, err, backoff := rpccs.relayInner(ctx, singleConsumerSession, relayResult, relayTimeout)
-	if err != nil {
-		failRelaySession := func(origErr error, backoff_ bool) {
-			backOffDuration := 0 * time.Second
-			if backoff_ {
-				backOffDuration = lavasession.BACKOFF_TIME_ON_FAILURE
+			// Extract fields from the sessionInfo
+			singleConsumerSession := sessionInfo.Session
+			epoch := sessionInfo.Epoch
+			reportedProviders := sessionInfo.ReportedProviders
+
+			relayRequest, err := lavaprotocol.ConstructRelayRequest(goroutineCtx, privKey, lavaChainID, chainID, &localRelayRequestData, providerPublicAddress, singleConsumerSession, int64(epoch), reportedProviders)
+			if err != nil {
+				responses <- &relayResponse{
```

Comment:
this pattern repeats itself on every return, we can use defer to do it once instead, i owuld use uniquely named variables like errRet instead of err to make sure scoping isn't overwritten

Reply:
Sure
Fixed 2241c0f5da1055375ef4b4d418224fd7cd1270b1